Concave activity jp

John-Polo opened the pull request:
I wrote an activity/code that uses r.param.scale to calculate vertical concavity on elev_lid792_1m.

John-Polo committed:
added a test function that does nothing called test_fun

John-Polo committed:
Create config_johnpolo.json

John-Polo committed:
Delete activity_john_polo

John-Polo committed:
Create activity_johnpolo.py

This is to calculate the profc: profile curvature (curvature intersecting with the plane defined by Z axis and maximum gradient direction). It is based on "Additional" from https://grasswiki.osgeo.org/wiki/Using_GRASS_GIS_through_Python_and_tangible_interfaces_(workshop_at_FOSS4G_NA_2016)
Positive values describe convex profile curvature, negative values concave profile curvature.

John-Polo committed:
Delete config_anna_petrasova.json

John-Polo commented:
I read the errors from the P.R. checks and it looks like I had a "," in the JSON that didn't need to be there. Same for an "env" in the .py. Removed them both.

John-Polo committed:
Update activity_johnpolo.py

John-Polo committed:
Update config_johnpolo.json

wenzeslaus commented:
> I had a "," in the JSON that didn't need to be there.

Yes, no trailing commas in JSON (while, confusingly, they need to be present in some cases in Python code formatted with Black).

> Same for an "env" in the .py. Removed them both.

The env is needed because that's how Tangible Landscape manages custom computational region settings within the GRASS session. You need to add it as a parameter to your function.

John-Polo committed:
Update activity_johnpolo.py

John-Polo commented:
It complained about env and I took env out. Now it complains that env is missing.

wenzeslaus reviewed (changes requested):
This looks good, but you still need to revert removal of the files created by Anna - we want them to stay in the code. I think the following should give you the files back:

```
git checkout upstream/main -- activities/activity_anna_petrasova.py
git checkout upstream/main -- activities/config_anna_petrasova.json
```

The `upstream` part of `upstream/main` requires that you already set up your repository to receive updates from the main (aka upstream) repository (and possibly re-created your fork since you used main branch before there - unrelated to issue at hand but otherwise relevant). You can skip all that and instead of `upstream/main` use `45855021947a32b041ae65cbc0e012a647f694a7` which is the hash of a revision which still contains Anna's files (and it happens to be also the one where she added them, but that's just a coincidence).

wenzeslaus commented on activities/activity_johnpolo.py:
```diff
@@ -0,0 +1,37 @@
+    elevation = "elev_lid792_1m"
+    select_meth = 'profc'
+    
+    run_topo_par(scanned_elev=elevation, morphmap='feat_verticalcurv', env=env, select_meth=select_meth)
```
It is perhaps unexpected - perhaps we oversimplified the template - but no custom parameters are allowed for these functions. They are called from Tangible Landscape which does not know what to put there, so these values need to be hardcoded in the function.

To avoid hardcoding at some level, you can create another function (_not_ starting with `run_`) without hardcoded values and call it from `run_topo_par` where you have these hardcoded values. This is little more complicated, so that's why I mentioned the (over)simplification in the template.

wenzeslaus commented on activities/config_johnpolo.json:
```diff
@@ -0,0 +1,16 @@
+      {
+      "layers": [
+        [
+          "d.rast", "map=elev_lid792_1m"
```
You probably want `feat_verticalcurv` here. Since it will be hardcoded in the `run_topo_par` function, it will exist when Tangible Landscape runs this function.

Then you will see this in the artifact from Render activities check.

John-Polo replied:
I substituted feat_verticalcurv for elev...

wenzeslaus replied:
Great. I checked the result in the artifact and it looks good!

wenzeslaus reviewed (changes requested):
Ha. I meant to request changes, not approve. Sorry for the confusing. But it is in a good shape.

wenzeslaus commented:
About the formatting complains from the Super-Linter (`Found errors in [black] linter!`), you don't need to fix them manually. Install Black and run it on the file. (The Black does not need to be integrated with GRASS GIS in any way. Black in any Python editor, in Anacodna, or [online](https://black.now.sh/?version=stable) will work.)

John-Polo committed:
Update activity_johnpolo.py

Get rid of whitespace

John-Polo committed:
Create activity_anna_petrasova.py

John-Polo committed:
Create config_anna_petrasova.json

John-Polo committed:
Update config_johnpolo.json

replace elev with feat_verticalcurv

John-Polo committed:
Update activity_anna_petrasova.py

John-Polo committed:
Update activity_johnpolo.py

John-Polo committed:
Update activity_johnpolo.py

I used the Black online. It inserted a blank last line, but I think there isn't supposed to be one?

John-Polo commented:
> 
> 
> About the formatting complains from the Super-Linter (`Found errors in [black] linter!`), you don't need to fix them manually. Install Black and run it on the file. (The Black does not need to be integrated with GRASS GIS in any way. Black in any Python editor, in Anacodna, or [online](https://black.now.sh/?version=stable) will work.)

Hi Vaclav,
I used the Black online that you linked to and I copied everything but the trailing blank line over to activities_johnpolo.py, so I should pass Super-Linter, but it still fails. When I look in the Super-Linter report it says that all of my files passed. I don't know what it's complaining about.

wenzeslaus commented on activities/activity_anna_petrasova.py:
```diff
@@ -2,15 +2,12 @@
 
 """
 Instructions
-
 - Functions intended to run for each scan
   need to be named run_xxxxx
-
 - Do not modify the parameters of the run_xxx function
   unless you know what you are doing
   see optional parameters:
   https://github.com/tangible-landscape/grass-tangible-landscape/wiki/Running-analyses-and-developing-workflows#python-workflows
-
```
Can you add back the empty lines here. Then this should disappear from the PR.

John-Polo replied:
I added the lines back in.

wenzeslaus commented:
> ...should pass Super-Linter, but it still fails. When I look in the Super-Linter report it says that all of my files passed.

The output is long, not ideal, so you need to scroll to find things. It is e.g. this line:

```
               return_code, 0, msg=("Running {filename} failed".format(**locals())),
```

But the issue looks like something which happened to me before when I used subsequently different versions of Black, but that's actually not that important. Important is that trailing comma in a line like the one above is taken by current release of Black as an instruction to spread the parameters over multiple lines. It is the default settings and I'm little confused about how exactly it works, but deleting that last (trailing) comma and running Black again should give you code which passes the check.

John-Polo committed:
delete trailing commas

I added blank lines in the instructions at top to separate each step. I deleted any "," that appeared right before ")" in activity_johnpolo.py (but it looks like activity_anna_petrasova.py has those marks) and I deleted a trailing empty line in activity_anna_petrasova.py.

John-Polo committed:
Update activity_anna_petrasova.py

Add new line at end.

John-Polo committed:
Update activity_anna_petrasova.py

Fix whitespace in trailing line.

John-Polo committed:
Update activity_johnpolo.py

commas!

John-Polo commented:
This is from the Super-Linter report: 

2021-02-24 14:42:05 [INFO]   File:[/github/workspace/activities/activity_luke_allen.py]
2021-02-24 14:42:05 [ERROR]   Found errors in [black] linter!
2021-02-24 14:42:05 [ERROR]   Error code: 1. Command output:
------
would reformat /github/workspace/activities/activity_luke_allen.py
Oh no! 💥 💔 💥

2021-02-24 14:42:05 [INFO]   File:[/github/workspace/tests/functions.py]
2021-02-24 14:42:06 [ERROR]   Found errors in [black] linter!
2021-02-24 14:42:06 [ERROR]   Error code: 1. Command output:
------
would reformat /github/workspace/tests/functions.py
Oh no! 💥 💔 💥

2021-02-24 14:42:07 [INFO]   File:[/github/workspace/website/render_activities.py]
2021-02-24 14:42:07 [ERROR]   Found errors in [black] linter!
2021-02-24 14:42:07 [ERROR]   Error code: 1. Command output:
------
would reformat /github/workspace/website/render_activities.py
Oh no! 💥 💔 💥

I have no way to fix the last 2, because I don't even know what they are and I'm not going to touch Luke's work. The report says activity_johnpolo.py passed, as well as config_johnpolo.json. This Super-Linter is annoying. :)

wenzeslaus commented on activities/activity_johnpolo.py:
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python3
+
+"""
+Instructions
+- Functions intended to run for each scan
+  need to be named run_xxxxx
+
+- Do not modify the parameters of the run_xxx function
+  unless you know what you are doing
+  see optional parameters:
+  https://github.com/tangible-landscape/grass-tangible-landscape/wiki/Running-analyses-and-developing-workflows#python-workflows
+
+- All gs.run_command/read_command/write_command/parse_command
+  need to be passed env parameter (..., env=env)
```
You can delete these instructions and replace them which once sentence description of what you are doing in the script.

See the second point before you delete the text - it relates to my other comment.

John-Polo replied:
I deleted the three statements and wrote a one sentence description. 

In response to the missing execution on the anna_petrasova.py file, I did the following in Git Bash:
john@DESKTOP-7OSMQHM MINGW64 /d/projects_and_data/data/landscape_git/gis714-2021-tangible-landscape (concave_activity_jp)
$ git update-index --chmod=+x activities/activity_anna_petrasova.py

It returned no errors, so I think it worked.

wenzeslaus commented on activities/activity_johnpolo.py:
```diff
@@ -0,0 +1,52 @@
+        output=morphmap,
+        method=select_meth,
+        env=env,
+        overwrite=True,
```
The `overwrite=True,` is actually not needed here because you set that globally with `env["GRASS_OVERWRITE"] = "1"`.

John-Polo replied:
I deleted `"overwrite=True"`.

wenzeslaus commented on activities/activity_johnpolo.py:
```diff
@@ -0,0 +1,52 @@
+import grass.script as gs
+
+
+def run_topo_par(scanned_elev, morphmap, select_meth, env, **kwargs):
```
The parameters of this function need to be the same as in the template. See e.g. Anna's script.

John-Polo replied:
I deleted `morphmap` and `select_meth` from `run_topo_par`.

wenzeslaus commented on activities/activity_johnpolo.py:
```diff
@@ -0,0 +1,52 @@
+
+    run_topo_par(
+        scanned_elev=elevation,
+        morphmap="feat_verticalcurv",
+        env=env,
+        select_meth=select_meth,
```
`select_meth` and `morphmap` need to be hardcoded in the `run_topo_par` function. In general, this would be a bad design, but here it is actually required by the Tangible Landscape API (which wouldn't know what to assign to these parameters).

John-Polo replied:
I edited the `gs.run.command` by adding: `output="feat_verticalcurv",` and `method="profc",`. This is how I understood what you mean by hardcoding the function. Hopefully, I got it right this time. :)

wenzeslaus replied:
Yes. That's it!

wenzeslaus reviewed (changes requested):
Just the issue with parameters of the `run_...` function remains from the main issues. See the comments.

For some reason restoring Anna's file did not restore the executable permissions which were set on the file. Although they are not needed in Tangible Landscape or in the continuous integrtaion, but let's assume Anna was making use of them and let's try to restore them. The command to do that is (if that doesn't work, let's just ignore this):

```
git update-index --chmod=+x activities/activity_anna_petrasova.py 
```

wenzeslaus commented:
> I have no way to fix the last 2, because I don't even know what they are and I'm not going to touch Luke's work.

Good call. You are not touching those with you commit, so there is no reason to try to fix them. The problem must be (is) elsewhere.

> The report says activity_johnpolo.py passed, as well as config_johnpolo.json.

As long as these two pass, we can merge this PR.

> This Super-Linter is annoying. :)

I'm actually little frustrated, because I can't figure out where is the issue, specifically why it surfaced only now. But that's fine because it is _not_ related to this PR and I made some changes in #17 to hopefully mitigate the issue.

As far as the specific linter goes - the Python code formatter Black - the idea is that you actually don't need to care about formatting much and you get consistent formatting in the whole project. There are issues and annoyances especially when enforcing it automatically (and Black is actually still in beta so there are even breaking changes here and there), but from my experience, the advantages outweigh that. For one script maybe not, for a project with multiple contributors definitively yes because you don't have to deal with different formatting styles etc. The assumption here is also that eventually, more time is spend reading the code than writing it, so spending some extra time with writing (e.g., in order to minimize distractions when reading it) should pay off in the long run. I hope this helps a little in dealing with these side issues.

John-Polo committed:
Vaclav described steps I need to take.

I deleted the Instructions, gave a brief description, removed parameters from my run_xxxx, hardcoded some other parameters, copied over some more code from the template, and changed permissions on a file via Git Bash.

John-Polo committed:
Take out resampled

Take out resampled

John-Polo committed:
Update activity_johnpolo.py

John-Polo commented:
Yaayy!!! :D

wenzeslaus commented:
> In response to the missing execution on the anna_petrasova.py file, ... It returned no errors, so I think it worked.

I don't see it would change in any commits. I will try myself just before merging it with the same command to see how it behaves.

wenzeslaus reviewed (approved):
This is ready to be merged. Congrats!

wenzeslaus commented:
Great. Close #10 as I assume this fully replaces it. I will merge shortly.

wenzeslaus committed:
Restore the original executable permissions

John-Polo commented:
I closed #10. 
I think. ;)
Am I supposed to close this one (#16) too? There is an option here that says "Close with comment".

wenzeslaus commented:
> > In response to the missing execution on the anna_petrasova.py file, ... It returned no errors, so I think it worked.
> 
> I don't see it would change in any commits. I will try myself just before merging it with the same command to see how it behaves.

OK, what worked for me was:

```
git update-index --chmod=+x activities/activity_anna_petrasova.py
git commit -m "Restore the original executable permissions"
git restore activities/activity_anna_petrasova.py
```

Notably, the `git commit` is without filename, `-a`, nor it is preceded by `git add`. Note to myself: The `git update-index` change happens in the staging area (it says `update-index`, so I should have expected that). On Linux, there was no actual change of permissions of the filename, so `git add` would revert the permissions back to the original state. The `git restore` then changes the actual permissions.

The file seems to be finally untouched by this commit in the *Files changed*. All the changes to the file will disappear when we do   _**Squash** and merge_.

wenzeslaus commented:
> I closed #10.

Thanks!

> Am I supposed to close this one (#16) too? There is an option here that says "Close with comment".

Not saying that you want to further work on it and having review with an approval usually indicates that it is ready to be merged (immediately or perhaps after final round of checks finishes). Merging needs to be done by project maintainers (with write access). Merging than closes the PR (closed as _merged_ as opposed to _closed_ as not to be further worked on - perhaps not the best terminology :-).
